applications/FluidDynamicsApplication/custom_processes/calulate_levelset_consistent_nodal_gradient_process.cpp

```diff
@@ -0,0 +1,148 @@
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Mohammad R. Hashemi
+//
+
+// Project includes
+#include "includes/define.h"
+#include "utilities/parallel_utilities.h"
+
+// Application includes
+#include "fluid_dynamics_application_variables.h"
+#include "calulate_levelset_consistent_nodal_gradient_process.h"
+
+
+namespace Kratos
+{
+
+/* Public functions *******************************************************/
+
+/// constructor
+CalulateLevelsetConsistentNodalGradientProcess::CalulateLevelsetConsistentNodalGradientProcess(
+        ModelPart& rModelPart)
+    : Process(), mrModelPart(rModelPart) {}
+
+/// Constructor with Kratos parameters.
+CalulateLevelsetConsistentNodalGradientProcess::CalulateLevelsetConsistentNodalGradientProcess(
+    ModelPart& rModelPart,
+    Parameters Parameters)
+    : CalulateLevelsetConsistentNodalGradientProcess(
+        rModelPart
+    ){}
+
+/// Constructor with Kratos model
+CalulateLevelsetConsistentNodalGradientProcess::CalulateLevelsetConsistentNodalGradientProcess(
+    Model& rModel,
+    Parameters Parameters)
+    : CalulateLevelsetConsistentNodalGradientProcess(
+        rModel.GetModelPart(Parameters["model_part_name"].GetString())
+    ){}
+
+void CalulateLevelsetConsistentNodalGradientProcess::Execute(){
+
+    KRATOS_TRY;
+
+    const auto zero_vector = ZeroVector(3);
+    // Set to zero
+    block_for_each(mrModelPart.Nodes(), [&](Node<3>& rNode){
+        rNode.SetValue(NODAL_AREA, 0.0);
+        rNode.SetValue(PRESSURE_GRADIENT, zero_vector);
+    });
+
+    // Current domain size
+    const unsigned int num_dim = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+    const unsigned int num_nodes = num_dim + 1; //For tetrahedra and triangles
+
+    // Auxiliar containers
+    double detJ0 = 0.0;
+    Matrix DN_DX, J0, InvJ0;
+    Vector N(num_nodes), pressures(num_nodes), distances(num_nodes), grad(num_dim);
+
+    // First element iterator
+    const auto it_element_begin = mrModelPart.ElementsBegin();
+
+    const int elements_number = mrModelPart.Elements().size();
+
+    // Iterate over the elements
+    #pragma omp parallel for firstprivate(DN_DX, N, J0, InvJ0, detJ0, pressures, distances, grad)
+    for(int i_elem = 0; i_elem < elements_number; ++i_elem) {
+
+        const auto it_elem = it_element_begin + i_elem;
+        auto& r_geometry = it_elem->GetGeometry();
+
+        // Current geometry information
+        const unsigned number_of_nodes = r_geometry.PointsNumber();
+
+        for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node){
+            distances(i_node) = r_geometry[i_node].FastGetSolutionStepValue(DISTANCE);
+        }
+
+        unsigned int nneg=0, npos=0;
+        for(unsigned int i = 0; i < number_of_nodes; ++i)
+        {
+            if(distances(i) > 0) {
+                npos += 1;
+            } else {
+                nneg += 1;
+            }
+        }
+
+        if(nneg == 0 || npos == 0)
```

Comment:
```suggestion
        if(!IsSplit)
```
and encapsulate the `nneg` and `npos` calculation and check in a private method.

Reply:
Done.